File: clients/typescript/src/transports/local.ts
import { type ChildProcess, spawn } from "node:child_process";
import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { createServer } from "node:net";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { PawrlyError } from "../errors.js";
import type { QueryHandle } from "../query.js";
import type {
  CacheEntryInfo,
  CatalogSnapshot,
  FunctionDescription,
  FunctionInfo,
  HealthReport,
  MaterializeOutcome,
  MaterializeSpec,
  RefreshCatalogOutcome,
  RefreshOutcome,
  ReloadReport,
  SemanticModelDescription,
  SemanticModelInfo,
  SemanticQuery,
  SourceDef,
  SourceInfo,
  SourceTestReport,
  TableDescription,
  TableInfo,
  VacuumReport,
} from "../result.js";
import type { Transport } from "../transport.js";
import { RestTransport } from "./rest.js";

export interface LocalOptions {
  config?: string;
  home?: string;
  binary?: string;
}

/** Spawns a `pawrly console` child on a private loopback port and talks to it
 * over REST — an engine this client owns and tears down. */
export class LocalTransport implements Transport {
  readonly name = "local";

  private constructor(
    private readonly rest: RestTransport,
    private readonly proc: ChildProcess,
    private readonly tmp: string,
  ) {}

  static async create(opts?: LocalOptions): Promise<LocalTransport> {
    const binary = opts?.binary ?? "pawrly";
    const tmp = mkdtempSync(join(tmpdir(), "pawrly-local-"));
    let config = opts?.config;
    if (!config) {
      config = join(tmp, "pawrly.yaml");
      writeFileSync(config, "version: 1\n");
    }
    const port = await freePort();
    const args: string[] = [];
    if (opts?.home) args.push("--home", opts.home);
    args.push("--config", config, "console", "--addr", `127.0.0.1:${port}`);
    const proc = spawn(binary, args, { stdio: ["ignore", "ignore", "pipe"] });

    const rest = new RestTransport(`http://127.0.0.1:${port}`);
    const self = new LocalTransport(rest, proc, tmp);

    const deadline = Date.now() + 10_000;
    for (;;) {
      if (proc.exitCode !== null) {
        self.cleanup();
        throw new PawrlyError("PAWRLY_INTERNAL", `pawrly console exited (${proc.exitCode})`);
      }
      try {
        if ((await rest.health()).ok) break;
      } catch {
        /* not up yet */
      }
      if (Date.now() > deadline) {
        self.terminate();
        self.cleanup();
        throw new PawrlyError("PAWRLY_INTERNAL", "pawrly console never became healthy");
      }
      await new Promise((r) => setTimeout(r, 50));
    }
    return self;
  }

  query(sql: string, params: Record<string, string>, limit?: number): Promise<QueryHandle> {
    return this.rest.query(sql, params, limit);
  }
  semanticQuery(q: SemanticQuery): Promise<QueryHandle> {
    return this.rest.semanticQuery(q);
  }
  explain(sql: string, analyze: boolean): Promise<string> {
    return this.rest.explain(sql, analyze);
  }
  cancel(queryId: string): Promise<boolean> {
    return this.rest.cancel(queryId);
  }
  materialize(name: string, spec: MaterializeSpec): Promise<MaterializeOutcome> {
    return this.rest.materialize(name, spec);
  }
  listSources(): Promise<SourceInfo[]> {
    return this.rest.listSources();
  }
  listTables(source?: string, nameGlob?: string): Promise<TableInfo[]> {
    return this.rest.listTables(source, nameGlob);
  }
  describeTable(name: string): Promise<TableDescription> {
    return this.rest.describeTable(name);
  }
  schemaSnapshot(sources?: string[], compact?: boolean): Promise<CatalogSnapshot> {
    return this.rest.schemaSnapshot(sources, compact);
  }
  cacheEntries(): Promise<CacheEntryInfo[]> {
    return this.rest.cacheEntries();
  }
  listFunctions(): Promise<FunctionInfo[]> {
    return this.rest.listFunctions();
  }
  describeFunction(namespace: string, name: string): Promise<FunctionDescription> {
    return this.rest.describeFunction(namespace, name);
  }
  listSemanticModels(): Promise<SemanticModelInfo[]> {
    return this.rest.listSemanticModels();
  }
  describeSemanticModel(name: string): Promise<SemanticModelDescription> {
    return this.rest.describeSemanticModel(name);
  }
  addSource(def: SourceDef): Promise<SourceInfo> {
    return this.rest.addSource(def);
  }
  removeSource(name: string): Promise<boolean> {
    return this.rest.removeSource(name);
  }
  testSource(name: string): Promise<SourceTestReport> {
    return this.rest.testSource(name);
  }
  reloadConfig(): Promise<ReloadReport> {
    return this.rest.reloadConfig();
  }
  refreshCatalog(source?: string): Promise<RefreshCatalogOutcome> {
    return this.rest.refreshCatalog(source);
  }
  refreshTable(name: string): Promise<RefreshOutcome> {
    return this.rest.refreshTable(name);
  }
  invalidateCache(name: string): Promise<boolean> {
    return this.rest.invalidateCache(name);
  }
  vacuumCache(): Promise<VacuumReport> {
    return this.rest.vacuumCache();
  }
  dropMaterialized(name: string): Promise<boolean> {
    return this.rest.dropMaterialized(name);
  }
  health(): Promise<HealthReport> {
    return this.rest.health();
  }
  shutdown(): Promise<void> {
    return this.rest.shutdown();
  }

  private terminate(): void {
    if (this.proc.exitCode === null) this.proc.kill("SIGTERM");
  }
  private cleanup(): void {
    rmSync(this.tmp, { recursive: true, force: true });
  }
  close(): void {
    this.rest.close();
    this.terminate();
    this.cleanup();
  }
}

function freePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const srv = createServer();
    srv.on("error", reject);
    srv.listen(0, "127.0.0.1", () => {
      const addr = srv.address();
      const port = addr && typeof addr === "object" ? addr.port : 0;
      srv.close(() => resolve(port));
    });
  });
}

File: clients/python/src/pawrly/transports/local.py
"""In-process transport: spawns a `pawrly console` child on a private loopback
port and talks to it over REST — an engine this client owns and tears down."""

import os
import shutil
import socket
import subprocess
import tempfile
import time

from ..errors import PawrlyError
from .rest import RestTransport


def _free_port() -> int:
    s = socket.socket()
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()
    return port


class LocalTransport(RestTransport):
    """A `pawrly console` the client spawns, owns, and tears down on `close()`."""

    name = "local"

    def __init__(
        self,
        config: str | None = None,
        home: str | None = None,
        binary: str = "pawrly",
    ) -> None:
        self._tmp = tempfile.mkdtemp(prefix="pawrly-local-")
        if config is None:
            config = os.path.join(self._tmp, "pawrly.yaml")
            with open(config, "w") as f:
                f.write("version: 1\n")

        port = _free_port()
        args = [binary]
        if home:
            args += ["--home", str(home)]
        args += ["--config", str(config), "console", "--addr", f"127.0.0.1:{port}"]
        self._proc = subprocess.Popen(
            args, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE
        )
        super().__init__(f"http://127.0.0.1:{port}")

        deadline = time.time() + 10
        while True:
            if self._proc.poll() is not None:
                err = self._proc.stderr.read().decode() if self._proc.stderr else ""
                self._cleanup()
                raise PawrlyError("PAWRLY_INTERNAL", f"pawrly console exited: {err}")
            try:
                if self.health().ok:
                    break
            except Exception:
                pass  # not up yet
            if time.time() > deadline:
                self._terminate()
                self._cleanup()
                raise PawrlyError(
                    "PAWRLY_INTERNAL", "pawrly console never became healthy"
                )
            time.sleep(0.05)

    def _terminate(self) -> None:
        if self._proc.poll() is None:
            self._proc.terminate()
            try:
                self._proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._proc.kill()

    def _cleanup(self) -> None:
        shutil.rmtree(self._tmp, ignore_errors=True)

    def close(self) -> None:
        super().close()
        self._terminate()
        self._cleanup()

    def __del__(self) -> None:
        try:
            self._terminate()
        except Exception:
            pass
